File: python/new_problems/reverse_linked_list.py
"""
Reverse Linked List

Problem: Reverse a singly linked list in-place.

Input: 1 → 2 → 3 → 4 → 5 → None
Output: 5 → 4 → 3 → 2 → 1 → None

Approach:
- Iterative with three pointers: prev, curr, next
- Reverse the direction of each link
- Time: O(n), Space: O(1)
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def reverseList(head):
    """
    Reverse a singly linked list.

    Args:
        head: Head of the linked list

    Returns:
        New head (previously the tail)

    Time: O(n)
    Space: O(1)
    """
    prev = None
    curr = head

    while curr:
        # Store next node before we change curr.next
        next_temp = curr.next
        # Reverse the link
        curr.next = prev
        # Move prev and curr one step forward
        prev = curr
        curr = next_temp

    return prev

def _build_list(values):
    """Helper: build linked list from list of values."""
    if not values:
        return None
    head = ListNode(values[0])
    curr = head
    for val in values[1:]:
        curr.next = ListNode(val)
        curr = curr.next
    return head

def _list_to_array(head):
    """Helper: convert linked list to array."""
    result = []
    curr = head
    while curr:
        result.append(curr.val)
        curr = curr.next
    return result

if __name__ == "__main__":
    # Test case 1: Normal list
    head1 = _build_list([1, 2, 3, 4, 5])
    result1 = reverseList(head1)
    print(f"Test 1: {_list_to_array(result1)} == [5, 4, 3, 2, 1]")

    # Test case 2: Single element
    head2 = _build_list([1])
    result2 = reverseList(head2)
    print(f"Test 2: {_list_to_array(result2)} == [1]")

    # Test case 3: Empty list
    head3 = _build_list([])
    result3 = reverseList(head3)
    print(f"Test 3: {_list_to_array(result3)} == []")

    # Test case 4: Two elements
    head4 = _build_list([1, 2])
    result4 = reverseList(head4)
    print(f"Test 4: {_list_to_array(result4)} == [2, 1]")

File: java/new_problems/ReverseLinkedList.java
/**
 * Reverse Linked List
 *
 * Problem: Reverse a singly linked list in-place.
 *
 * Input: 1 → 2 → 3 → 4 → 5 → null
 * Output: 5 → 4 → 3 → 2 → 1 → null
 *
 * Approach:
 * - Iterative with three pointers: prev, curr, next
 * - Reverse the direction of each link
 * - Time: O(n), Space: O(1)
 */
public class ReverseLinkedList {

    public static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
    }

    /**
     * Reverse a singly linked list.
     *
     * @param head Head of the linked list
     * @return New head (previously the tail)
     *
     * Time: O(n)
     * Space: O(1)
     */
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;

        while (curr != null) {
            // Store next node before we change curr.next
            ListNode nextTemp = curr.next;
            // Reverse the link
            curr.next = prev;
            // Move prev and curr one step forward
            prev = curr;
            curr = nextTemp;
        }

        return prev;
    }

    private static ListNode buildList(int[] values) {
        if (values.length == 0) return null;
        ListNode head = new ListNode(values[0]);
        ListNode curr = head;
        for (int i = 1; i < values.length; i++) {
            curr.next = new ListNode(values[i]);
            curr = curr.next;
        }
        return head;
    }

    private static String listToString(ListNode head) {
        StringBuilder sb = new StringBuilder();
        ListNode curr = head;
        while (curr != null) {
            sb.append(curr.val).append(" → ");
            curr = curr.next;
        }
        sb.append("null");
        return sb.toString();
    }

    public static void main(String[] args) {
        ReverseLinkedList solver = new ReverseLinkedList();

        // Test case 1: Normal list
        ListNode head1 = buildList(new int[]{1, 2, 3, 4, 5});
        ListNode result1 = solver.reverseList(head1);
        System.out.println("Test 1: " + listToString(result1));

        // Test case 2: Single element
        ListNode head2 = buildList(new int[]{1});
        ListNode result2 = solver.reverseList(head2);
        System.out.println("Test 2: " + listToString(result2));

        // Test case 3: Empty list
        ListNode head3 = buildList(new int[]{});
        ListNode result3 = solver.reverseList(head3);
        System.out.println("Test 3: " + listToString(result3));

        // Test case 4: Two elements
        ListNode head4 = buildList(new int[]{1, 2});
        ListNode result4 = solver.reverseList(head4);
        System.out.println("Test 4: " + listToString(result4));
    }
}
